In [3]:
import pandas as pd

df = pd.read_csv('Data/Telco-Customer-Churn.csv')
print(f"Original shape: {df.shape}")

# Simulate new data arriving
new_rows = df.head(100).copy()
df_updated = pd.concat([df, new_rows], ignore_index=True)
df_updated.to_csv('Data/Telco-Customer-Churn.csv', index=False)
print(f"Updated shape: {df_updated.shape}")

Original shape: (7043, 21)
Updated shape: (7143, 21)


## Summary

**What I built:** DVC data versioning — track datasets like Git 
tracks code, restore any previous version instantly.

**Key concepts:**
- DVC stores actual data in remote storage (S3, GCS, local)
- Git stores only the .dvc pointer file (tiny, contains MD5 hash)
- MD5 hash fingerprints exact file contents — one row change = new hash
- dvc add → creates pointer file, moves data to DVC cache
- dvc push → uploads data to remote storage
- dvc checkout → restores exact dataset for current Git commit

**The workflow:**
1. dvc add dataset.csv → creates dataset.csv.dvc
2. git commit dataset.csv.dvc → versions the pointer
3. Data changes → dvc add again → new hash → new git commit
4. git checkout old_commit dataset.csv.dvc → dvc checkout → original data restored

**Why this matters:**
- Git alone: can't version large files, breaks on CSVs > 100MB
- DVC: handles any size, stores in cloud, restores in seconds
- Every experiment is reproducible — same code + same data = same model

**Original dataset:** md5=0f9de68e012bd3aed5fa7cdc9fc421af (7043 rows)
**Updated dataset:** md5=06c4cf1cf83e1f52432bc56f8d079e6a (7143 rows)
**Restored original:** Shape (7043, 21) ✅

**In production replace:**
/tmp/dvc_storage → s3://your-bucket/dvc-storage

**Next:** A/B testing framework — Day 48